src/foam/web/DetachedURLState.js

```diff
@@ -0,0 +1,212 @@
+          deserialization.`,
+      factory: function() {
+        return this.Parser.create({
+          strict: true,
+          creationContext: this.__subContext__
+        });
+      }
+    },
+    {
+      name: 'bindingsMap_',
+      documentation: 'Map of {<key>: foam.core.Slot} comprising state.',
+      factory: function() { return {}; }
+    },
+    {
+      name: 'subMap_',
+      documentation: `Map of {<key>: <event subscription>} parallel to
+          bindingsMap_.`,
+      factory: function() { return {}; }
+    },
+    {
+      name: 'unboundMap_',
+      documentation: `Map of {<key>: <deserialized value>} that contains
+          bindings loaded from hash, but not bound to a foam.core.Slot.`,
+      factory: function() { return {}; }
+    },
+    {
+      class: 'String',
+      name: 'path_',
+      documentation: `"path" part of hash; hash is:
+          "#<path>?<key1>=<value1>&<key2>=<value2>..."`,
+      postSet: function(old, nu) {
+        if ( old !== nu ) this.onStateChange();
+      }
+    },
+    {
+      class: 'String',
+      name: 'hash_',
+      documentation: 'Complete state serialized to a URL hash value.',
+    },
+  ],
+
+  methods: [
+    function init() {
+      this.SUPER();
+      this.hashToState_();
+    },
+    function getHash() { return this.hash_; },
+    function setHash(hash) {
+      this.hash_ = hash;
+      this.hashToState_();
+    },
+    function getPath() { return this.path_; },
+    function setPath(path) { return this.path_ = path; },
+    function getSlot(name) {
+      return this.bindingsMap_[name] || null;
+    },
+    function addBinding(name, slot) {
+      if ( this.bindingsMap_.hasOwnProperty(name) ) {
+        this.warn('Overwriting URLState:', name, this.bindingsMap_[name],
+                  'with', slot);
+        this.subMap_[name].detach();
+      }
+      if ( this.unboundMap_.hasOwnProperty(name) ) {
+        slot.set(this.unboundMap_[name]);
+        delete this.unboundMap_[name];
+      }
+      this.bindingsMap_[name] = slot;
+      this.subMap_[name] = slot.sub(this.onStateChange);
+      this.onStateChange();
+    },
+    function removeBinding(name) {
+      if ( ! this.bindingsMap_.hasOwnProperty(name) ) return;
+      this.subMap_[name].detach();
+      delete this.subMap_[name];
+      delete this.bindingsMap_[name];
+      this.onStateChange();
+    },
+    function clearBindings() {
+      var subMap = this.subMap_;
+      for ( var key in subMap ) {
+        if ( ! subMap.hasOwnProperty(key) ) continue;
+        subMap[key].detach();
+      }
+      this.subMap_ = {};
+      this.bindingsMap_ = {};
+      this.unboundMap_ = {};
+      this.onStateChange();
+    },
+    function stateToHash_() {
+      var hash = '#' + this.path_ + '?';
+      var bindingsMap = this.bindingsMap_;
+      var first = true;
+      function out(str) { hash += str; }
+
+      // Output bindings, then unbound values.
+      this.outputMapToHashStr_(
```

Comment:
I feel like this would be easier to follow if the helper accepted a current string and appended key/value pairs to it. You can replace the first flag with an empty string check, and return the string at the end instead of the flag.

Reply:
Done. @arobins PTAL